src/hyperlight_host/src/hypervisor/kvm.rs

```diff
@@ -55,6 +65,527 @@ pub(crate) fn is_hypervisor_present() -> bool {
+        /// The maximum number of supported breakpoints is `Self::MAX_NO_OF_HW_BP`
+        pub fn set_breakpoints(&mut self, vcpu_fd: &VcpuFd, step: bool) -> Result<()> {
+            let addrs: &[u64] = &self.hw_breakpoints;
+
+            if addrs.len() >= Self::MAX_NO_OF_HW_BP {
+                return Err(new_error!(
+                    "The breakpoints array is larger than expected: {} >= {}",
+                    addrs.len(),
+                    Self::MAX_NO_OF_HW_BP
+                ));
+            }
+
+            self.dbg_cfg.arch.debugreg = [0; 8];
+            for (k, addr) in addrs.iter().enumerate() {
+                self.dbg_cfg.arch.debugreg[k] = *addr;
+                self.dbg_cfg.arch.debugreg[7] |= 1 << (k * 2);
+            }
+
+            if !addrs.is_empty() {
+                self.dbg_cfg.control |= KVM_GUESTDBG_USE_HW_BP;
+            } else {
+                self.dbg_cfg.control &= !KVM_GUESTDBG_USE_HW_BP;
+            }
+
+            if step {
+                self.dbg_cfg.control |= KVM_GUESTDBG_SINGLESTEP;
+            } else {
+                self.dbg_cfg.control &= !KVM_GUESTDBG_SINGLESTEP;
+            }
+
+            log::debug!("Setting bp: {:?} cfg: {:?}", addrs, self.dbg_cfg);
+            vcpu_fd
+                .set_guest_debug(&self.dbg_cfg)
+                .map_err(|e| new_error!("Could not set guest debug: {:?}", e))?;
+
+            self.single_step = step;
+
+            Ok(())
+        }
+    }
+
+    impl KVMDriver {
+        /// Resets the debug information to disable debugging
+        fn disable_debug(&mut self) -> Result<()> {
+            self.debug = Some(KvmDebug::default());
+
+            self.set_single_step(false)
+        }
+
+        /// Returns the instruction pointer from the stopped vCPU
+        pub fn get_instruction_pointer(&self) -> Result<u64> {
+            let regs = self
+                .vcpu_fd
+                .get_regs()
+                .map_err(|e| new_error!("Could not retrieve registers from vCPU: {:?}", e))?;
+
+            Ok(regs.rip)
+        }
+
+        /// Sets or clears stepping for vCPU
+        pub fn set_single_step(&mut self, enable: bool) -> Result<()> {
+            let debug = self
+                .debug
+                .as_mut()
+                .ok_or_else(|| new_error!("Debug is not enabled"))?;
+
+            debug.set_breakpoints(&self.vcpu_fd, enable)
+        }
+
+        /// Gdb expects the target to be stopped when connected.
+        /// This method provides a way to set a breakpoint at the entrypoint
+        /// it does not keep this breakpoint set after the vcpu already stopped at the address
+        pub fn set_entrypoint_bp(&self) -> Result<()> {
+            if self.debug.is_some() {
+                log::debug!("Setting entrypoint bp {:X}", self.entrypoint);
+                let mut entrypoint_debug = KvmDebug::new();
+                entrypoint_debug.hw_breakpoints.push(self.entrypoint);
```

Comment:
Could we make an method on KvmDebug for adding a breakpoint that does error checking (>4 breapoints), instead of manually modifying the vec? It seems fragile to do the error checking during the `set_breakpoints`, and to expose the inner vec like this.

Reply:
good point, I'll change it 👍